crates/cashu/src/nuts/nut02.rs

```diff
@@ -103,26 +123,110 @@ impl Id {
 
     /// [`Id`] from bytes
     pub fn from_bytes(bytes: &[u8]) -> Result<Self, Error> {
-        Ok(Self {
-            version: KeySetVersion::from_byte(&bytes[0])?,
-            id: bytes[1..].try_into()?,
-        })
+        let version = KeySetVersion::from_byte(&bytes[0])?;
+        let id = match version {
+            KeySetVersion::Version00 => IdBytes::V1(bytes[1..].try_into()?),
+            KeySetVersion::Version01 => IdBytes::V2(bytes[1..].try_into()?),
+        };
+        Ok(Self { version, id })
     }
 
+    pub fn get_version(&self) -> KeySetVersion {
+        self.version
+    }
+
+    /*
     /// [`Id`] as bytes
     pub fn as_bytes(&self) -> [u8; Self::BYTELEN + 1] {
         let mut bytes = [0u8; Self::BYTELEN + 1];
         bytes[0] = self.version.to_byte();
-        bytes[1..].copy_from_slice(&self.id);
+        match self.id {
+            IdBytes::V1(id) => bytes[1..].copy_from_slice(&id),
+            IdBytes::V2(id) => bytes[1..].copy_from_slice(&id),
+        }
         bytes
+    }*/
+
+    /// create [`Id`] v2 from keys, unit and (optionally) expiry
+    /// 1 - sort public keys by their amount in ascending order
+    /// 2 - concatenate all public keys to one byte array
+    /// 3 - concatenate the unit string to the byte array (e.g. "unit:sat")
+    /// 4 - If a final expiration is specified, convert it into a radix-10 string and concatenate it (e.g "final_expiry:1896187313")
+    /// 5 - HASH_SHA256 the concatenated byte array and take the first 31 bytes
+    /// 6 - prefix it with a keyset ID version byte
+    pub fn v2_from_data(map: &Keys, unit: &CurrencyUnit, expiry: Option<u64>) -> Self {
+        let mut keys: Vec<(&Amount, &super::PublicKey)> = map.iter().collect();
+        keys.sort_by_key(|(amt, _v)| *amt);
+
+        let mut pubkeys_concat: Vec<u8> = keys
+            .iter()
+            .map(|(_, pubkey)| pubkey.to_bytes())
+            .collect::<Vec<[u8; 33]>>()
+            .concat();
+
+        // Add the unit
+        pubkeys_concat.extend(b"unit:");
+        pubkeys_concat.extend(unit.to_string().as_bytes());
```

Comment:
Unit should be converted to lowercase before hashing: https://github.com/cashubtc/nuts/pull/182#discussion_r1939111715